move/axelar_gateway/sources/versioned/gateway_v0.move

```diff
@@ -0,0 +1,480 @@
+        message_id,
+    );
+
+    self[command_id] == message_status::executed()
+}
+
+/// To execute a message, the relayer will call `take_approved_message`
+/// to get the hot potato `ApprovedMessage` object, and then trigger the app's package via discovery.
+public(package) fun take_approved_message(
+    self: &mut GatewayV0,
+    source_chain: String,
+    message_id: String,
+    source_address: String,
+    destination_id: address,
+    payload: vector<u8>,
+): ApprovedMessage {
+    let command_id = message::message_to_command_id(source_chain, message_id);
+
+    let message = message::new(
+        source_chain,
+        message_id,
+        source_address,
+        destination_id,
+        bytes32::from_bytes(hash::keccak256(&payload)),
+    );
+
+    assert!(
+        self[command_id] == message_status::approved(message.hash()),
+        EMessageNotApproved,
+    );
+
+    let message_status_ref = &mut self[command_id];
+    *message_status_ref = message_status::executed();
+
+    events::emit_message_executed(
+        message,
+    );
+
+    // Friend only.
+    channel::create_approved_message(
+        source_chain,
+        message_id,
+        source_address,
+        destination_id,
+        payload,
+    )
+}
+
+public(package) fun send_message(_self: &GatewayV0, message: MessageTicket, current_version: u64) {
+    let (
+        source_id,
+        destination_chain,
+        destination_address,
+        payload,
+        version,
+    ) = message.destroy();
+    assert!(version <= current_version, ENewerMessage);
+    events::emit_contract_call(
+        source_id,
+        destination_chain,
+        destination_address,
+        payload,
+        address::from_bytes(hash::keccak256(&payload)),
+    );
+}
+
+// -----------------
+// Private Functions
+// -----------------
+
+fun peel_messages(message_data: vector<u8>): vector<Message> {
+    utils::peel!(
+        message_data,
+        |bcs| {
+            let messages = vector::tabulate!(
+                bcs.peel_vec_length(),
+                |_| message::peel(bcs),
+            );
+            assert!(messages.length() > 0, EZeroMessages);
+            messages
+        },
+    )
+}
+
+fun data_hash(command_type: u8, data: vector<u8>): Bytes32 {
+    let mut typed_data = vector::singleton(command_type);
+    typed_data.append(data);
+
+    bytes32::from_bytes(hash::keccak256(&typed_data))
+}
+
+fun approve_message(self: &mut GatewayV0, message: message::Message) {
+    let command_id = message.command_id();
+
+    // If the message was already approved, ignore it.
+    if (self.messages.contains(command_id)) {
+        return
+    };
+
+    self
+        .messages
+        .add(
+            command_id,
+            message_status::approved(message.hash()),
+        );
+
+    events::emit_message_approved(
+        message,
+    );
+}
+
+/// ---------
+/// Test Only
+/// ---------
+
+#[test_only]
+public fun operator(self: &GatewayV0): &address {
```

Comment:
```suggestion
public(package) fun operator(self: &GatewayV0): &address {
```
for sanity, let's only expose methods at the package level unless needed later, even in tests?